File: src/main/java/xyz/phanta/cbtweaker/recipe/ComponentSet.java
package xyz.phanta.cbtweaker.recipe;

import gnu.trove.impl.Constants;
import gnu.trove.map.TObjectIntMap;
import gnu.trove.map.hash.TObjectIntHashMap;

public class ComponentSet {

    public static final ComponentSet EMPTY = new ComponentSet() {
        @Override
        public void put(String componentId, int count) {
            throw new UnsupportedOperationException();
        }
    };

    private final TObjectIntMap<String> components = new TObjectIntHashMap<>(
            Constants.DEFAULT_CAPACITY, Constants.DEFAULT_LOAD_FACTOR, 0);

    public void put(String componentId) {
        put(componentId, 1);
    }

    public void put(String componentId, int count) {
        if (count > 0) {
            components.adjustOrPutValue(componentId, count, count);
        }
    }

    public void addFrom(ComponentSet o) {
        o.components.forEachEntry((c, n) -> {
            put(c, n);
            return true;
        });
    }

    public int getCount(String componentId) {
        return components.get(componentId);
    }

    public ComponentSet copy() {
        ComponentSet result = new ComponentSet();
        result.addFrom(this);
        return result;
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/util/machine/ComponentSet.kt
package st.evening.mc.cbtweaker.util.machine

import gnu.trove.impl.Constants
import gnu.trove.iterator.TObjectIntIterator
import gnu.trove.map.TObjectIntMap
import gnu.trove.map.hash.TObjectIntHashMap
import st.evening.mc.cbtweaker.util.TObjectIntEmptyIterator

interface ComponentSet {
    fun getCount(componentId: String): Int

    fun iterator(): TObjectIntIterator<String>

    object Empty : ComponentSet {
        override fun getCount(componentId: String): Int = 0

        override fun iterator(): TObjectIntIterator<String> = TObjectIntEmptyIterator()
    }
}

class MutableComponentSet : ComponentSet {
    companion object {
        fun copyOf(o: ComponentSet): MutableComponentSet = MutableComponentSet().also { it.addAll(o) }
    }

    private val components: TObjectIntMap<String> = TObjectIntHashMap(
        Constants.DEFAULT_CAPACITY, Constants.DEFAULT_LOAD_FACTOR, 0
    )

    fun put(componentId: String, count: Int) {
        if (count > 0) {
            components.adjustOrPutValue(componentId, count, count)
        }
    }

    fun put(componentId: String) {
        put(componentId, 1)
    }

    fun addAll(o: ComponentSet) {
        val iter = o.iterator()
        while (iter.hasNext()) {
            iter.advance()
            put(iter.key(), iter.value())
        }
    }

    override fun getCount(componentId: String): Int = components.get(componentId)

    override fun iterator(): TObjectIntIterator<String> = components.iterator()
}
